### ISSUE TITLE
Root file finding mechanism fails when \input uses macros

### ISSUE THREAD

**Issue by p4jo** [James-Yu/LaTeX-Workshop#2351]
Magic comments declaring a main file are cool, but I want as much of the work as possible to be automated by the root file finding algorithm - i.e. not having to add a magic comment to every included file. 

When I have a file (g.tex) that has a magic comment to its root file (b.tex) and several files (content1.tex,...) included in it (via \input), then the root file finding algorithm should find the correct root file (b) when the editor is in any of the included files (content1,...). The magic comment in g.tex should incentivise the algorithm to mark the file g and every file in g as included in b.

The problem with the root finding algorithm:

The LaTeX Workshop stopped finding the root files and started always outputting "Found files that might be root, choose the first one: something.tex, b.tex, ..." and choosing the first one. The correct root file based on the active editor is indeed one of those, but it doesn't choose the right one. It tries to parse a lot of .fls files but misses the associated .fls files of the "might be root" files [they are all in the workspaceRootDir, have no spaces in file names, are not excluded from root finding and have the (superfluous) "\begin{document}"]. The associated .fls files do actually contain all the content files that get input so if it would parse them it should be able to find the root file based on the active file.

The suggestion - i.e. helping the root finding algorithm by inserting magic comments - arises as a workaround. 

My setup: b.tex has "\newcommand{\pathtog}{g.tex}\input{Setup.tex}\begin{document}\end{document}". Setup includes something like "\begin{document}\input{\pathtog}\end{document}". Parsing Setup is not helpful and it is therefore excluded from the root finding mechanism. g.tex has "% !TeX root = b.tex" and all the "\input{content1.tex}"....

There are several such b and g files, the gs and Setup are in (seperate) subfolders and the paths and names are just placeholders and actually make sense in the real case.

**Comment by jlelong** [comment]
Please provide a minimal working example to reproduce your problem and post the full logs of the extension.

**Comment by p4jo** [comment]
[TeX - MWP.zip](https://github.com/James-Yu/LaTeX-Workshop/files/5453765/TeX.-.MWP.zip)


Here is a minimal working example.

**Comment by jlelong** [comment]
As you pointed out, the correct root file cannot be found using the parser because the `\input` commands use variables. The only way to detect the root file correctly in such cases is to rely on the `.fls` file. 

Surprisingly, we only use the `.fls` file once the root file is found. We should also use the `.fls` files to discover the correct root file.

**Comment by p4jo** [comment]
Yes, that's it. Of course it's not "standard" to include files like that. However, the reason, why I seemed to expect that it does work in my original post, is that it had worked for some time and broke as I renamed the g files, at least as far as I remember.

The suggestion with the magic comments is another story, but becomes redundant in most cases when the root file detection via  fls files works. Of course there could be even more magic comments like marking files as potential root files without having \begin{document} or marking files with \begin{document} as non-root (Workshop starts parsing files that are excluded from root finding if they are the active editor), but those are different issues.
Thanks for maintaining this project!

**Comment by p4jo** [comment]
Hello, I have to come back to this now that I have some time again.
I changed my setup such that no macros, but rather real paths are used in the \input{} commands for my content files.
It still shows `Found files that might be root, choose the first one:` way too often. It even shows it immediately after switching from the root file to an included file (via two \input{} commands) although the file was added to file watcher. 
I couldn't easily reproduce it using a minimal working example.
This is actually annoying. One way to prevent it may be removing this "feature" `Found files that might be root, choose the first one:` and staying with the last root file.
